### PR TITLE
Encapsulate itype (part 1)

### PR DESCRIPTION
**Pull request by mugling**
`itype` can be considered an implementation detail of `item`. This PR starts a series of changes to improve encapsulation

### CODE REVIEW

File: src/npc.cpp [hunk 1]
```diff
@@ -1387,7 +1387,7 @@ int npc::player_danger( const player &ur ) const
    ret += 8;
  } else if( u->weapon_value( u->weapon ) > 20 )
   ret++;
- else if (u->weapon.type->id == "null") // Unarmed
```
Comment (mugling) [inline src/npc.cpp]: we want to avoid constructs such as this

Comment (Coolthulhu) [inline src/npc.cpp]: We have `player::is_armed()` that does this.

Comment (mugling) [inline src/npc.cpp]: even better

File: src/iexamine.cpp [hunk 2]
```diff
@@ -70,11 +70,10 @@ void iexamine::gaspump(player &p, const tripoint &examp)
...
             if( one_in(10 + p.get_dex()) ) {
                 add_msg(m_bad, _("You accidentally spill the %s."), item_it->type_name().c_str());
-                item spill( item_it->typeId(), calendar::turn );
-                const int spill_min = item_it->liquid_charges( 1 );
                 ///\EFFECT_DEX decreases amount of gas spilled from a pump
-                const int spill_max = item_it->liquid_charges( 1 ) * 8.0 / std::max( 1, p.get_dex() );
-                spill.charges = rng( spill_min, spill_max );
+                const int qty = rng( item_it->liquid_charges( 1 ),
+                                     item_it->liquid_charges( 1 ) * 8.0 / std::max( 1, p.get_dex() ) );
+                item spill( item_it->type, calendar::turn, qty );
```
Comment (codemime) [inline src/iexamine.cpp]: Why not `item::split( int qty )`? It prevents both exposing `item::type` on explicit construction and reassignment of non-encapsulated `item::charges` (here and below).

File: src/iexamine.cpp [hunk 3]
```diff
@@ -1559,7 +1558,7 @@ std::list<item> iexamine::get_harvest_items( const itype &type, const int plant_
         return result;
     }
     const islot_seed &seed_data = *type.seed;
-    const itype_id &seed_type = type.id;
+    const itype_id &seed_type = item( &type ).typeId();
```
Comment (Coolthulhu) [inline src/iexamine.cpp]: Creating an entire item of a type just to get the ID of said type is really ugly.

**Comment by Coolthulhu** [comment]
Creating an entire item just to get the ID of its type is a really ugly hack. It would be better to create a static method, like `static itype_id &item::get_id( itype &type )` that would handle getting ID of a type without creating an entire item for it.

File: src/iexamine.cpp (hunk 3 above)
Comment (mugling) [inline src/iexamine.cpp]: Yes, very ugly, however `get_harvest_items` needs refactoring and that would remove the problem

Comment (codemime) [inline src/iexamine.cpp]: Can be resolved nicely via moving `get_harvest_items(...)` into `item` (where it probably belongs).

**Comment by codemime** [comment]
@Coolthulhu It looks like a hack too. If we pass `itype` into the static function, we already have access to the said `itype`. We either have broken encapsulation or we access `itype::id` directly in this case. Otherwise we use `item:ItemId()`. BTW Why CamelCase?

**Comment by mugling** [comment]
Needing to determine the `id` of a bare `itype` is a code smell and usually indicative of a design problem. I'd prefer to avoid the static member as it similarly breaks encapsulation.

I'd prefer to leave the hack in-situ for now and refactor later but perhaps you disagree?

File: src/player.cpp [hunk 4]
```diff
@@ -13436,7 +13436,7 @@ std::vector<std::string> player::get_overlay_ids() const
 
     // last weapon
     // TODO: might there be clothing that covers the weapon?
-    if( !weapon.is_null() ) {
+    if( !is_armed() ) {
```
Comment (BevapDin) [inline src/player.cpp]: `is_armed` returns `! is_null()`, now you test for `!! weapon.is_null()`

**Comment by Coolthulhu** [comment]
> I'd prefer to leave the hack in-situ for now and refactor later but perhaps you disagree?

The hack isn't any better than directly accessing the field. I'd rather delay encapsulation until it can be done properly.

File: src/player.cpp (hunk 4 above)
Comment (mugling) [inline src/player.cpp]: is that necessarily bad?

Comment (codemime) [inline src/player.cpp]: It inverts the logic?

Comment (mugling) [inline src/player.cpp]: ah sorry yes I see

File: src/iexamine.cpp (hunk 2 above)
Comment (mugling) [inline src/iexamine.cpp]: Good suggestion

**Comment by Coolthulhu** [comment]
Another argument in favor of not using the item creation hack: an explicit hack method (for example, a static getter) could be easily removed after everything is fixed, while constructor invocation will require actually reading the code to check if it is an instance of the hack or actually useful constructor invocation.

**Comment by mugling** [comment]
@Coolthulhu that could work although the other option would be making classes `friend` of `itype` and gradually removing those as soon as practical. Has the added advantage that it's self-documenting

**Comment by kevingranade** [comment]
A static getter does less violence to encapsulation, declaring other
classes as friends subverts everything useful about encapsulation.

**Comment by mugling** [comment]
> A static getter does less violence to encapsulation, declaring other classes as friends subverts everything useful about encapsulation.

Absolutely true, however at present there is no encapsulation of `itype` to subvert. Opinion seems to favor the method however so I'll go with that route and add documentation marking it as deprecated

**Comment by kevingranade** [comment]
I added an accessor for id on itype, there's no reason for it to be a static, it's just an accessor that's deprecated, but some things need to be refactored before it can be eliminated.

**Comment by kevingranade** [comment]
Oh right, once the accessor use is restricted to item.cpp, we could define a function in item.cpp that looks it up and is friended by itype.  I think that'll work anyway, and maximally discourages new things using it.

**Pull request merged**